cc-wrapper: Allow user to override -target for clang

mmlb opened the pull request:
###### Description of changes

Only sets -target in wrapped compiler if none is provided by the user. This is especially helpful for generating bpf objects.

Fixes #176128

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mmlb committed:
cc-wrapper: Allow for override of -target for clang/clang++

This enables users to make use of clang's multi-platform/target support
without having to go through full cross system setup. This is especially useful
for generating bpf object files, I'm not even usre what would a no-userland
cross compile system tuple even look like to even try going that route.

Fixes #176128

mmlb commented:
ping @Mindavi for review too

mmlb commented:
I couldn't find anything in the repo that sets `-target` (as expected) which leads me to believe that its relatively safe.

mmlb commented:
@06kellyjac give this a try

Ericson2314 commented:
Also I don't think this is a good fix for https://github.com/NixOS/nixpkgs/issues/176128, because clang might try to produce a far binary for *both* arches instead. What you want is logic in cc-wrapper itself to only add a `-target ...` if the user didn't already pass one themselves.

mmlb commented:
> Also I don't think this is a good fix for #176128, because clang might try to produce a far binary for _both_ arches instead.

I find this surprising if true, got any links to documentation? So far everything I've seen to generate fat binaries or bundles is very much opt in with either clang args (clang offload bundler) or different tools (lipo, but I think this is osx only)

> What you want is logic in cc-wrapper itself to only add a `-target ...` if the user didn't already pass one themselves.

Setting aside the fat binaries, I think this makes sense anyway so will look into updating `cc-wrapper`.

mmlb committed:
cc-wrapper: Use case statements instead of bunch of if/elif checks

Makes for easier to read code imo, case-in-point there was a duplicate test for
`$p = -E` before. While doing this little bit of refactor I changed rest ->
kept because that makes more sense, rest sounds like its the rest of params
while kept says these are the params that we've kept.

I tested for no change in behavior using the following bash script:

```
reset() {
	cInclude=1
	cxxInclude=1
	cxxLibrary=1
	dontLink=0
	isCxx=0
	nonFlagArgs=0
	params=()
}

parseParams() {
	declare -i n=0
	nParams=${#params[@]}
	while (("$n" < "$nParams")); do
		p=${params[n]}
		p2=${params[n + 1]:-} # handle `p` being last one

		case "$p" in
		-[cSEM] | -MM) dontLink=1 ;;
		-cc1) cc1=1 ;;
		-nostdinc) cInclude=0 cxxInclude=0 ;;
		-nostdinc++) cxxInclude=0 ;;
		-nostdlib) cxxLibrary=0 ;;
		-x)
			case "$p2" in
			*-header) dontLink=1 ;;
			c++*) isCxx=1 ;;
			esac
			;;
		-?*) ;;
		*) nonFlagArgs=1 ;; # Includes a solitary dash (`-`) which signifies standard input; it is not a flag
		esac
		n+=1
	done
}

for p in c S E M MM; do
	reset
	params=-$p
	parseParams
	[[ $dontLink != 1 ]] && echo "expected dontLink=1 for params:${params[@]}" >&2 && exit 1
done

reset
params=(-x foo-header)
parseParams
[[ $dontLink != 1 ]] && echo "expected dontLink=1 for params:${params[@]}" >&2 && exit 1

reset
params=(-x c++-foo)
parseParams
[[ $isCxx != 1 ]] && echo "expected isCxx=1 for params:${params[@]}" >&2 && exit 1

reset
params=-nostdlib
parseParams
[[ $cxxLibrary != 0 ]] && echo "expected cxxLibrary=0 for params:${params[@]}" >&2 && exit 1

reset
params=-nostdinc
parseParams
[[ $cInclude != 0 ]] && echo "expected cInclude=0 for params:${params[@]}" >&2 && exit 1
[[ $cxxInclude != 0 ]] && echo "expected cxxInclude=0 for params:${params[@]}" >&2 && exit 1

reset
params=-nostdinc++
parseParams
[[ $cxxInclude != 0 ]] && echo "expected cxxInclude=0 for params:${params[@]}" >&2 && exit 1

reset
params=-cc1
parseParams
[[ $cc1 != 1 ]] && echo "expected cc1=1 for params:${params[@]}" >&2 && exit 1

reset
params=-
parseParams
[[ $nonFlagArgs != 1 ]] && echo "expected nonFlagArgs=1 for params:${params[@]}" >&2 && exit 1

reset
params=bleh
parseParams
[[ $nonFlagArgs != 1 ]] && echo "expected nonFlagArgs=1 for params:${params[@]}" >&2 && exit 1

reset
params=-?
parseParams
[[ $nonFlagArgs != 0 ]] && echo "expected nonFlagArgs=0 for params:${params[@]}" >&2 && exit 1

exit 0
```

mmlb commented:
Hey @Ericson2314 I've rebased on-top of staging and re-targted the PR accordingly. Also made it so -target should only show up once in the final params just once. This ends up causing a full rebuild of clang which then ripples out, this did not happen with the previous approach, just an fyi.

Mindavi commented:
I feel like there was a reason for the if, elif chaining, but maybe it's just how it was written.

Could that be because we're using a very old bash on MacOS and this is used during bootstrap or something? Hopefully someone else can confirm or deny.

@happysalada do you know?

mmlb commented:
> I feel like there was a reason for the if, elif chaining, but maybe it's just how it was written.
> 
> Could that be because we're using a very old bash on MacOS and this is used during bootstrap or something? Hopefully someone else can confirm or deny.
> 
> @happysalada do you know?

I thought it might have been due to sh vs bash, but I verified that dash supports case and so I'd suspect old bash would too.

mmlb commented:
`case` is in posix sh (see 2.9.4 in https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09) so should be ok to use for old mac bash.

Mindavi commented:
Thanks for checking! It does look nicer with case IMO, so that's good.

06kellyjac commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -170,6 +160,10 @@ if [ "$dontLink" != 1 ]; then
 fi
 
+if [ "$setClangTarget" == 1 ]; then
+    extraAfter+=(-target $clangTarget)
```
@Mindavi ^

mmlb replied:
:man_facepalming: great catch. I'll actually remove the old bit in the next push.

mmlb replied:
Removed the old code: https://github.com/NixOS/nixpkgs/pull/176152/files#diff-ad7b67d4a92a2db43281c9c45aa74c32341a554fbca0b6b25be593f4648abb5eL300-L307

Mindavi replied:
Nice, now it makes more sense :)

Ericson2314 commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -29,6 +29,8 @@ cc1=0
 cxxLibrary=1
 cInclude=1
+setClangTarget=0
+clangTarget=none
```
```suggestion
```

mmlb replied:
ack

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -216,6 +216,7 @@ stdenv.mkDerivation {
         export named_cxx=${targetPrefix}g++
       elif [ -e $ccPath/clang ]; then
         wrap ${targetPrefix}clang $wrapper $ccPath/clang
+        substituteInPlace $ccPath/clang --replace setClangTarget=0 setClangTarget=1 --replace clangTarget=none clangTarget=${targetPlatform.config}
```
Don't need
```suggestion
```

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -216,6 +216,7 @@ stdenv.mkDerivation {
         export named_cxx=${targetPrefix}g++
       elif [ -e $ccPath/clang ]; then
```
```suggestion
      elif [ -e $ccPath/clang ]; then
        export clangTarget=${targetPlatform.config}
```

mmlb replied:
ack

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -226,6 +227,7 @@ stdenv.mkDerivation {
         ln -s ${targetPrefix}g++ $out/bin/${targetPrefix}c++
       elif [ -e $ccPath/clang++ ]; then
         wrap ${targetPrefix}clang++ $wrapper $ccPath/clang++
+        substituteInPlace $ccPath/clang++ --replace setClangTarget=0 setClangTarget=1 --replace clangTarget=none clangTarget=${targetPlatform.config}
```
```suggestion
        export clangTarget=${targetPlatform.config}
        wrap ${targetPrefix}clang++ $wrapper $ccPath/clang++
```

mmlb replied:
ack

mmlb commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -170,6 +160,10 @@ if [ "$dontLink" != 1 ]; then
 fi
 
+if [ "$setClangTarget" == 1 ]; then
+    extraAfter+=(-target $clangTarget)
```
@Ericson2314 will empty @ variables also be substituted with empty string/nothing or do they get skipped over and left as is?

We still need some way to pass `setClangTarget` from default.nix into cc-wrapper.sh. I can either do something like `export setClangTarget=1` and then instead of of `if [ $setClangTarget...` I can do `if [ @isClang@`. But if unknown @ vars are left as empty then I can do something like:

```
clangTarget=@clanTarget@

...

case $p in
...
-target) clangTarget=$p2
...
esac

...

if [ -n $clangTarget ]; then
    extraAfter+=(-target $clangTarget)
fi
```

Ericson2314 replied:
I would call `setClangTarget` `needsTarget` and just it `0` or `1` based on the whether is clang in default.nix

Let's try to make sure all variables to substitute are defined to *something* at build time,

mmlb replied:
:+1:, also just read from https://nixos.org/manual/nixpkgs/stable/#fun-substituteAll that unknown @ vars are left as is so wouldn't work the way I laid it out in the snippet above.

mmlb replied:
Should be good now.

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -208,13 +208,16 @@ stdenv.mkDerivation {
         export named_cxx=${targetPrefix}g++
       elif [ -e $ccPath/clang ]; then
+        needsTarget=1
+        export clangTarget=${targetPlatform.config}
```
Cause the other rename I would call this `defaultTarget` too.

mmlb replied:
will fix

mmlb replied:
updated

Mindavi commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -38,35 +39,23 @@ nParams=${#params[@]}
+        -nostdinc) cInclude=0 cxxInclude=0 ;;
+        -nostdinc++) cxxInclude=0 ;;
+        -nostdlib) cxxLibrary=0 ;;
+        -target|--target=*) needsTarget=0;;
```
Is this automatically passed through?

mmlb replied:
Yep params comes from https://github.com/NixOS/nixpkgs/pull/176152/files#diff-5b214d4b19e3f1e9146a43659772fb2e7a5830badc02d53087ea62f5f5860b93R34
and params is maybe reset in the NIX_ENFORCE_PURITY if branch, but that basically filters out any -I/-L/-isystem args that `badPath` doesn't like from params. similar happens for march/mtune=native filtering.

So in the end ${params[@]} ends up with the params provided by the caller possibly without any impure paths or mtune/march flags

mmlb commented:
@Ericson2314 @Mindavi how do you all feel about this?

```diff
@@ -76,19 +76,13 @@ if [[ "${NIX_ENFORCE_PURITY:-}" = 1 && -n "$NIX_STORE" ]]; then
     while (( "$n" < "$nParams" )); do
         p=${params[n]}
         p2=${params[n+1]:-} # handle `p` being last one
-        if [ "${p:0:3}" = -L/ ] && badPath "${p:2}"; then
-            skip "${p:2}"
-        elif [ "$p" = -L ] && badPath "$p2"; then
-            n+=1; skip "$p2"
-        elif [ "${p:0:3}" = -I/ ] && badPath "${p:2}"; then
-            skip "${p:2}"
-        elif [ "$p" = -I ] && badPath "$p2"; then
-            n+=1; skip "$p2"
-        elif [ "$p" = -isystem ] && badPath "$p2"; then
-            n+=1; skip "$p2"
-        else
-            rest+=("$p")
-        fi
+
+        case "$p" in
+            -L/* | -I/*) badPath "${p:2}" && skip "${p:2}" && continue ;;
+            -L | -I | -isystem) badPath "$p2" && skip "$p2" && continue ;;
+        esac
+
+        rest+=("$p")
         n+=1
     done
     # Old bash empty array hack
```

I originally skipped over looking over this use of params because I didn't need to mess with it but when I went looking to make sure params will have -target to answer @Mindavi 's question I noticed that this too may look better as a case statement. Should I change this one too?

Ericson2314 commented:
@mmlb I like the sound of those. Just be careful with the `n+=1` I think you do need more of it as before we got (in affect) `n+=2` in certain cases.

mmlb commented:
> @mmlb I like the sound of those. Just be careful with the `n+=1` I think you do need more of it as before we got (in effect) `n+=2` in certain cases.

Yep, I totally missed that in the snippet. The actual commit has it though. I ended simplifying what I had in the snippet further too.

Ericson2314 commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -81,29 +70,29 @@ fi
+            -I | -L | -isystem) path=$p2 n+=1 ;;
+        esac
+
+        if badPath "$path"; then
```
```suggestion
        if test -N "$path" && badPath "$path"; then
```
I am not sure `badPath` should be happy with the empty path, an in any case I rather not rely on that.

mmlb replied:
badPath is happy with the empty case since it returns early if $path != /* basically. But yeah I can be explicit about it.

mmlb commented:
I'm looking into why glibc isn't building anymore :/

Ericson2314 commented:
Last thing from me, check out https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/bintools-wrapper/add-darwin-ldflags-before.sh from @thefloweringash. Perhaps we want to do the same thing for this? Then any compiler that doesn't need the stuff doesn't have the code in its wrapper.

I like that as a first step to modularizing the wrappers and prevent mass rebuilds in the future. (changing the Clang-only bits wouldn't cause a rebuild of things just using GCC.)

mmlb commented:
hmm I'm trying to build `gcc` and on staging I'm getting

```
nix-build --check -A gcc 2>&1 |tail
  /nix/store/wjbv1k6yigmb280wrvc1gkv8cnrsacij-glibc-2.34.tar.xz
  /nix/store/wp9v6jcv4vdxp8vwrfsb3q6qfalf2n5l-libunistring-1.0.tar.gz
  /nix/store/x4jhvxj104vdciivf1zvijncyd4rd3dp-mpc-1.2.1.tar.gz
  /nix/store/x9v0hmr76dm68pijzldzxks01iqzag1z-isl-0.20.tar.xz
  /nix/store/y3yiminrckvhf35fh9q42vjwi0npznji-acl-2.3.1.tar.gz
  /nix/store/ycj1q918l6852kndr75rlwmw43k5ybys-texinfo-6.8.tar.xz
  /nix/store/z1wfvm57izp9mdx3svdm8ayvw9f24bsr-bash51-002
  /nix/store/z4amhpapl83iwjxkq9p9lylkh0ffyq1d-bash51-013
  /nix/store/zpa3bjjp3awz7db16dl58cwi7b4jrl26-python-setup-hook.sh
error: some outputs of '/nix/store/s6sz5wka3cqgmbgkrvaw5pm72hik3d6h-gcc-wrapper-11.3.0.drv' are not valid, so checking is not possible
```

Mindavi commented:
Remove the --check, for starters?

mmlb commented:
> Last thing from me, check out https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/bintools-wrapper/add-darwin-ldflags-before.sh from @thefloweringash. Perhaps we want to do the same thing for this? Then any compiler that doesn't need the stuff doesn't have the code in its wrapper.
> 
> I like that as a first step to modularizing the wrappers and prevent mass rebuilds in the future. (changing the Clang-only bits wouldn't cause a rebuild of things just using GCC.)

Yeah I'll take a look at doing similar, agree that clang stuff shouldn't cause gcc to rebuild. Ideally the opposite would be true but making cc-wrapper and co agnostic would be more work than I'd want right now :smile: .

mmlb commented:
It *looks* like the fail is in the case that handles -I/, -I, -L/, -L, -isystem. My build is still going, but I'm pretty sure it would have failed by now. I'll wait until this build is done and then go about debugging why that case is failing. Of course its the one that was "obvious" and I didn't try testing ... :grin:

mmlb committed:
cc-wrapper: Add clang specific options to clang specific file

This way gcc doesn't need to be rebuilt because of clang. This also avoids
rebuilding clang when only the wrapper needs to be tweaked.

mmlb commented:
@Ericson2314 how's this look?

Ericson2314 commented:
I would tweak the names and make them consistent, but otherwise looks good!

mmlb commented:
> I would tweak the names and make them consistent, but otherwise looks good!

Can you elaborate? I copied the naming scheme used in bintools-wrapper:
`add-darwin-ldflags-before.sh` -> `add-clang-cflags-before.sh`
`add-local-ldflags-before.sh` -> `add-local-cflags-before.sh`

Do you want me to prefix `-cflags` with `cc`?

Mindavi commented:
Does this file already exist or is it missing: add-local-cflags-before.sh?

Ericson2314 commented:
@mmlib I don't know why bintools wrapper uses two different names either. Do you, @thefloweringash?

Ericson2314 commented:
I don't understand why the file in the Nixpkgs src has one name (with .sh) and the file "installed" as part of the built wrapper has a different name (still with .sh). I get it when we strip the .sh on install (e.g. installing something to `bin/`) but I don't get why for something very internal like this.

I also don't understand why anything is called "local". What does "local" mean here?

Finally, I don't like the use of "clang" because this applies equally well to any compiler that would accept `-target`. Just as we renamed the variables in bash to get rid of "clang" we should also rename the file to get rid of "clang".

mmlb commented:
ah gotcha. Yeah I think the idea is that there's an "interface" and clang and gcc can both implement it by having a file named `local-cc-cflags-before.sh` dropped into $out somewhere. I think that makes some sense. The other option would be instead of file names we'd source all files in a dir that match a pattern but that would be up to collation order and would be totally different than what is being done currently, and idk if its even warranted tbh.

I took `local` as an indicator that file has code to run vs just a list of values to be read, open to w/e name makes sense to show that distinction.

re clang vs more fine grained names. I can rename so its just for target, but some of the other files here are specific to the tool not to the features so I went with that too since only clang does -target and thus yagni.

Sounds like you want me to rename from clang -> target and maybe come up with something else more relevant than `local`. I'll do the former rn and maybe come up with a better name than local. We can discuss that separately.

Ericson2314 commented:
OK that makes some sense we can live with this for now.

mmlb commented:
> (the approval and some old junk queued up from before acidently)

what do you mean?

Mindavi commented:
John means that he had some comments in an 'open review' and they were submitted when he merged/approved. That wasn't the intention since they were stale.

Ericson2314 commented:
Yes exactly. Thanks.

mmlb commented:
ah, gotcha.
